mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveShell.java

```diff
@@ -0,0 +1,173 @@
+import org.apache.hive.service.cli.session.HiveSession;
+import org.apache.hive.service.server.HiveServer2;
+import org.apache.iceberg.common.DynMethods;
+
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+import static org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkState;
+
+/**
+ * Test class for running HiveQL queries, essentially acting like a Beeline shell in tests.
+ *
+ * It takes a metastore URL via conf, and spins up an HS2 instance which connects to it. The shell will only accept
+ * queries if it has been previously initialized via {@link #start()}, and a session has been opened via
+ * {@link #openSession()}. Prior to calling {@link #start()}, the shell should first be configured with props that apply
+ * across all test cases by calling {@link #setHiveConfValue(String, String)} ()}. On the other hand, session-level conf
+ * can be applied anytime via {@link #setHiveSessionValue(String, String)} ()}, once we've opened an active session.
+ */
+public class TestHiveShell {
+
+  private static final DynMethods.StaticMethod FIND_FREE_PORT = DynMethods.builder("findFreePort")
+          .impl("org.apache.hadoop.hive.metastore.utils.MetaStoreUtils")
+          .impl("org.apache.hadoop.hive.metastore.MetaStoreUtils")
+          .buildStatic();
+
+  private final HiveServer2 hs2;
+  private final HiveConf conf;
+  private CLIService client;
+  private HiveSession session;
+  private boolean started;
+
+  public TestHiveShell() {
+    conf = initializeConf();
+    hs2 = new HiveServer2();
+  }
+
+  public void setHiveConfValue(String key, String value) {
+    checkState(!started, "TestHiveShell has already been started. Cannot set Hive conf anymore.");
+    conf.set(key, value);
+  }
+
+  public void setHiveSessionValue(String key, String value) {
+    checkState(session != null, "There is no open session for setting variables.");
+    try {
+      session.getSessionConf().set(key, value);
+    } catch (Exception e) {
+      throw new RuntimeException("Unable to set Hive session variable: ", e);
+    }
+  }
+
+  public void start() {
+    checkState(isNotBlank(conf.get(HiveConf.ConfVars.METASTOREURIS.varname)),
+            "hive.metastore.uris must be supplied in config. TestHiveShell needs an external metastore to connect to.");
+    hs2.init(conf);
+    hs2.start();
+    client = hs2.getServices().stream()
+            .filter(CLIService.class::isInstance)
+            .findFirst()
+            .map(CLIService.class::cast)
+            .get();
+    started = true;
+  }
+
+  public void stop() {
+    if (client != null) {
+      client.stop();
+    }
+    hs2.stop();
+    started = false;
+  }
+
+  public void openSession() {
+    checkState(started, "You have to start TestHiveShell first, before opening a session.");
+    try {
+      SessionHandle sessionHandle = client.getSessionManager().openSession(
+              CLIService.SERVER_VERSION, "", "", "127.0.0.1", Collections.emptyMap());
```

Comment:
Nit: continuing indentation is off in this file

Reply:
fixed